### ISSUE TITLE
Authorization Token from Mobile App

### ISSUE THREAD

**Issue by ZicPL** [Piwigo/Piwigo#2168]
Hi
I want to say that I am very happy with using Piwigo. The author of the app deserves great **Applause**.

**Reason:**
Nowadays, security is playing an increasingly important role. The computing power used to crack passwords has gained significant efficiency. For this reason, a strong password **isn't as secure as it used to be**. Sorry for adding another topic, but I think I started an important problem. I guess that many users would like their private photos, videos to be well protected. Piwigo has a mobile app that is great for this feature 👏 

**Suggestion for marking the topic:**
Type: Feature
Type: Enhancement
Section: Users & Groups
Priority: High
Milestone: Yes

**Main part:**
Does the Product Owner plan to consider my request to create a 2FA (two-factor authentication) or MFA for the Piwigo app based on an event in a **trusted device**. A plugin for Piwigo doesn't seem like a good solution for anyone (I saw that one plugin exists in beta), but it is such a basic and integral feature that it shouldn't be implemented as an add-on.

**Proposed scenario for Product Owner:**
1. In the application, the Piwigo webapp Administrator can select an option (I treat the solution I propose as an option - the Administrator has the right not to decide on protection): "Login only with authorization in the Piwigo mobile app." for themselves or for a specific user/group.
2. The Administrator is shown the Token key in case of loss of the device where the mobile app was used for authorization.
3. Warning about data backup:
- Token Key
- Media (photos, videos, etc.) in case of server failure, loss of the token and the device with the mobile app.
4. The Administrator logs into the Piwigo mobile app and receives a message about adding the device to trusted. Even if someone steals the password, logging in to the new mobile app won't nothing. Trusted devices can be managed only via the Piwigo webapp (website sends requests to new devices - adding them to trusted). At this point, the captured password won't allow you to log in to the Piwigo application website, because it already verifies it through a trusted device with the Piwigo mobile app.
5. The User account created by the Administrator receives the following information upon login: "Request for secure login to <Piwigo instance name>. Install the mobile app and follow the steps given in the guide" - from now on, the user has 7 days to complete authorization. If they don't do it in this specified time, their account will be blocked. After blocking, they must write a message to the Administrator asking for unblocking.

Summary:
There are already 2 applications (webapp and mobile). All that remains is:
1. Creating a **push token mechanism (encrypted)** to the mobile application from the server <Piwigo instance name>.
2. Encryption with a salt of token keys in the DataBase

I'd appreciate a full answer. Please don't treat my message as a demand - it is a request.

Best regards
Luke

**Comment by GJSchaller** [comment]
I support this - I have a similar request on the Piwigo forums, but I agree TOTP should be a core feature:

https://piwigo.org/forum/viewtopic.php?id=33186

**Comment by ZicPL** [comment]
> I support this - I have a similar request on the Piwigo forums, but I agree TOTP should be a core feature:
> 
> https://piwigo.org/forum/viewtopic.php?id=33186

Thank You for Your interest 😍  Of course, Your topic also concerns security: OTP. I think that six-digit tokens should be mainly for administrator (expire every 30 seconds). Please keep in mind the different age of the users - for older people it is easier to click "accept" than to copy the numbers from another app.
My proposal seems easier and faster (and maybe even safer) to implement for everyone because the server and mobile app already exist. For the Administrator I'd consider your proposal (I use OTP in other services and I think it is a good way). Thank you for your understanding.

**Comment by alwanga** [comment]
I concur with this. i wanted to request for the same. 

At least MFA/2FA with Google Authenticator, Microsoft Authenticator or with the mobile app will be a super nice security improvement.

Thank you PIWIGO team!!

**Comment by alwanga** [comment]
@ZicPL i concur with your perspective. All we would need is a way to not only rely alone on username/password.
@Eric-Piwigo  @Piwigo-TranslationTeam , any thoughts about this issue, please?

**Comment by LintyDev** [comment]
Hello everyone,
From what I understand of the conversation, here are the key points (please let me know if I’ve missed anything):

- Strengthening Piwigo’s security through two-factor/multi-factor authentication (2FA/MFA).
- Offer this enhanced security via the Piwigo mobile application.
- Adding a 2FA/MFA option based on well-known TOTP providers (Google Authenticator, Microsoft Authenticator, etc.).
- Allowing these parameters to be configured according to specific user needs.
- Integrating this system directly into Piwigo’s core.

I’d also like to add another point:
In using the Piwigo API for an external project, I’m forced to store user credentials (username/password) to maintain a persistent session. This approach is not practical or safe. Implementing a more robust and API-friendly authentication system such as secure tokens would greatly simplify connection management and eliminate the need to store sensitive passwords.

What do you think?

**Comment by GJSchaller** [comment]
I need to say I would prefer TOTP over Mobile App.  TOTP codes can be backed up and restored, and are not dependent on a specific device / app - you can choose to use Google Authenticator, Authy, Duo, MS Authenticator, etc.  Also, this should be easier to implement than a dedicated connection to a specific app.

While I'm not against App-Specific MFA, I would prefer to see TOTP first, then other methods.

**Comment by RobyB89** [comment]
>  Good Luck 👋

@ZicPL don't leave.. You were the only one leading this issue in the right direction. If someone from Piwigo reacts (soon ?), it would be good if you were with us 😜

**Comment by LintyDev** [comment]
Today, we discussed this topic with @plegall , and here’s what came out of it: At Piwigo, we have already had a similar topic, which was about passkeys, brought up by @EddyLB . A solution that could replace passwords and is more secure, but we are also aware that this system is still a bit too early for the world.

For the implementation in Piwigo: In the first stage, we could have a two-factor authentication (2FA) plugin included in the Piwigo build (similar to LocalFilesEditor or AdminTools), with a 2FA functionality via third-party apps (such as Authy, 1Password, Bitwarden, etc.), with the option to enforce or not enforce the use of 2FA for users. 

**Why as a plugin?** Because releasing a new plugin is easier and faster than releasing a new version of Piwigo, giving us time to develop all the features. 

**Why only 2FA via third-party apps initially?** Because, in the beginning, it’s simpler and more secure to integrate into Piwigo. These changes also require us to rethink how we use the API (I will create another issue related to this). Personally, I believe that adding functionalities (different types of 2FA) gradually is preferable as it allows us to focus on making each feature robust. This gradual approach also allows us to better address user needs and ensures that the security features are both effective and user-friendly.

In the second phase, we could add this functionality through the iOS/Android Piwigo apps. This is where we return to @ZicPL’s original request, and I want to emphasize your argument about older people or those without technical skills. It could be reassuring to have the entire authentication system within the "official" Piwigo environment. While third-party apps like Google Authenticator and Authy are great, the simplicity of receiving push notifications directly within the Piwigo app could make the process much easier to adopt for users who aren’t as familiar with technology.

In all cases, these functionalities will follow the TOTP convention (Time-based One-Time Password), which means the use of a unique secret key per user, generating a time-sensitive code (e.g., 123456) that is valid for a short time (usually 30 seconds). This is a widely accepted standard for 2FA and ensures both security and compatibility with many third-party apps ( @GJSchaller , for the use of TOTP method, I agree with you )

@ZicPL , I think the configuration in your screen should actually be in the gallery (in different way) so that the user can choose their preferred 2FA method?

**Comment by alwanga** [comment]
Thank you for your contribution, @LintyDev,

I am glad to hear that there is a plan for traction on this issue.

I am not a PHP developer, but I am thinking of joining the Piwigo dev community to contribute to the implementation of these ideas.

In the meantime, we appreciate all the support we have received and will receive from the Piwigo team, esteemed contributors in this chat, and the community.

**Comment by LintyDev** [comment]
Hello @ZicPL :D
Hello @alwanga :D

> I am not a PHP developer, but I am thinking of joining the Piwigo dev community to contribute to the implementation of these ideas.

It's always a pleasure to see the community growing, you can also join the forum https://piwigo.org/forum/

> What do You think I should change title of this thread -> "Authorization : 2FA + Token from Mobile App"? Maybe I'm being unnecessarily stubborn about limiting the discussion to one of authorization methods. I see that this thread is treated in general about these two security techniques. @LintyDev Let me know if You think this is a good idea, ok?

Yes, it's a good idea, because in the code, these methods will work together to a certain point! Above all, we're aiming for user comfort by offering several solutions for secure access to piwigo.

> That's true. I use open-source alternative to Google Authenticator. It's really cool and I recommend it to everyone. Unfortunately, You won't convince me that it will be faster than accepting a notification from mobile app e.g. number of steps for login acceptance in Piwigo mobile app is much lower than TOTP.

To be honest, at the moment it's difficult to release new versions of the Piwigo NG (Android) mobile application. But we're also working on.

> I absolutely do not prioritize the implementation of this feature. This is a free, high-level discussion in which I like culture and rationalism. My reason I created this topic was just one thing. Piwigo mobile app already exists. It's a huge asset. If mobile app didn't exist, I wouldn't waste anyone's time (cost of both would be too high). Moreover, it's a very high level of security and combined with a trusted device, it will be as safe as mobile apps of Banks (world's top - yes yes I know that bank apps have audits, but this solution is so close.). Piwigo would certainly gain positive publicity among security experts. It's really worth taking this thread seriously.

I understand, and we are taking this topic seriously. In my last response, I refocused the discussion on 2FA, particularly via third-party apps. This initial approach allows us to quickly integrate a robust and secure feature while adhering to current standards. However, we are considering both methods: 2FA via third-party apps and MFA via the Piwigo apps. 

> I know I didn't try my best. It was just a quick example. Would You mind uploading here a screenshot, where could be these two security features?

I haven’t told Alice, our designer, about the project yet, so please go easy on me 😁
I’ve started working on a plugin to integrate all these methods. I began with the 2FA method ( @ZicPL you’re probably going to get tired of me, haha sorry). Here’s how it could look (it’s already functional).

- Add a new parameter to profile.php
![Capture d’écran 2024-12-17 à 13 32 08](https://github.com/user-attachments/assets/88f32209-a05b-4930-a9da-36db1153524b)

- The `Activate` button will open a modal window with the 2FA activation procedure. 
![Capture d’écran 2024-12-17 à 13 35 05](https://github.com/user-attachments/assets/e2e18234-477e-46eb-bdb7-91f470db810a)

At each connection, if the user has activated the 2FA, there will be a field to enter his totp code.

**Comment by LintyDev** [comment]
Hello @x-g3niu5 

The topic is still ongoing. Within the team, we're discussing releasing it alongside the next major update (Piwigo 16.0.0) because we want to integrate this plugin into Piwigo's build. This also ties into a discussion about adding an API key, as the implementation of two-factor authentication could cause some issues with the current API.

**Comment by LintyDev** [comment]
@GJSchaller Hi, currently on the master branch of the repo the version is `16.0.0beta1`. It is the 15.x branch that you will have the latest versions of 15

@TesterEU Hi, as I said in the thread, of course it will happen soon. And for admin control I've set the base. You can reset the 2FA on each user for example (they will have to initialise it again). But it's true that there will be more control by adding authentication via the piwigo mobile app.
